### DISCUSSION TITLE
Improved environment variable naming for collections in static configuration

### DISCUSSION THREAD

**Discussion by sandord** [dotnet/yarp#893] [Ideas] [answered]
The way configuration collections such as Routes are currently mapped is as a list:

```json
    "ReverseProxy": {
      "Routes": [
        {
          "RouteId": "WebsiteRoute",
          "Match": {
            "Hosts": [ "localhost:4200" ]
          }
        },
        {
          "RouteId": "Route2"
        }
      ]
    }
```

However, when overriding one of its underlying values with an environment variable, we'll have to name it as such:

```
    ReverseProxy__Routes__0__Hosts=localhost:4300
```

I know this is a limitation of [EnvironmentVariablesConfigurationProvider](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.configuration.environmentvariables.environmentvariablesconfigurationprovider?view=dotnet-plat-ext-5.0), which currently offers no way to index an array or list by any other way than an integer position.

It would be way less error-prone however, and a lot clearer if we could just do this:

```
    ReverseProxy__Routes__WebsiteRoute__Hosts=localhost:4300
```

I'm thinking of a way to provide an alternative mapping, without breaking the way it currently works.

Perhaps there's a way to provide the following mapping as an alternative, where the name of each direct descendant of `Routes` implies a `RouteId`?

```json
    "ReverseProxy": {
      "Routes": {
        "WebsiteRoute": {
          "Match": {
            "Hosts": [ "localhost:4200" ]
          }
        },
        "Route2": {
        }
      }
    }
```

This would then allow the use of the `ReverseProxy__Routes__WebsiteRoute__Hosts` environment variable name.

**Comment by Tratcher** [comment, marked as answer]
https://github.com/microsoft/reverse-proxy/issues/279 would address this if we decide to change the routes collection from an array to named objects.

In the meantime take a look at this new sample:
https://github.com/microsoft/reverse-proxy/pull/883/files#diff-9592b5ea7e975364e9ef73e6cf06df66ef4e4d08fdf16740fd9054169843ae79

**Reply by sandord** [reply]
Great, the way that sample provides environment variable substitution is really nice. Thanks!

**Reply by MihaZupan** [reply]
@sandord FYI #279 has been resolved by #900 and is included in the [latest preview (11)](https://github.com/microsoft/reverse-proxy/releases/tag/v1.0.0-preview11)
